support/src/main/java/org/jenkinsci/plugins/workflow/support/steps/MilestoneStepExecution.java

```diff
@@ -0,0 +1,356 @@
+        }
+        return new XmlFile(new File(j.getRootDir(), MilestoneStep.class.getName() + ".xml"));
+    }
+
+    @Extension
+    public static final class Listener extends RunListener<Run<?,?>> {
+        @Override public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (!(r instanceof FlowExecutionOwner.Executable) || ((FlowExecutionOwner.Executable) r).asFlowExecutionOwner() == null) {
+                return;
+            }
+            exit(r);
+        }
+    }
+
+    private static final class Milestone {
+
+        /**
+         * Numbers of builds currently running in this milestone.
+         */
+        final Set<Integer> holding = new TreeSet<Integer>();
+
+        /**
+         * Maximum permitted size of {@link #holding}, or null for unbounded.
+         */
+        @CheckForNull
+        Integer concurrency;
+
+        /**
+         * Context of the build currently waiting to enter this stage, if any.
+         */
+        @CheckForNull StepContext waitingContext;
+
+        /**
+         * Number of the build corresponding to {@link #waitingContext}, if any.
+         */
+        @Nullable
+        Integer waitingBuild;
+
+        /**
+         * Externalizable ID of the build corresponding to {@link #waitingContext}, if any.
+         */
+        @Nullable
+        String waitingBuildExternalizableId;
+
+        /**
+         * Last build that passed through the milestone, or null if none passed yet.
+         */
+        @CheckForNull
+        Integer lastBuild;
+
+        /**
+         * Last build extenalizable ID hat passed through the milestone, or null if none passed yet.
+         */
+        @CheckForNull
+        String lastBuildExternalizableId;
+
+        @Override public String toString() {
+            return "Stage[holding=" + holding + ", waitingBuild=" + waitingBuild + ", concurrency=" + concurrency + "]";
+        }
+
+        /**
+         * Unblocks the build currently waiting.
+         *
+         * @param message a message to print to the log of the unblocked build
+         */
+        void unblock(String message) {
+            assert Thread.holdsLock(MilestoneStepExecution.class);
+            assert waitingContext != null;
+            assert waitingBuild != null;
+            assert !holding.contains(waitingBuild);
+            println(waitingContext, message);
+            waitingContext.onSuccess(null);
+            holding.add(waitingBuild);
+            lastBuild = waitingBuild;
+            lastBuildExternalizableId = waitingBuildExternalizableId;
+            waitingContext = null;
+            waitingBuild = null;
+            waitingBuildExternalizableId = null;
+        }
+    }
+
+    /**
+     * Records that a flow was canceled while waiting in a stage step because a newer flow entered that stage instead.
+     */
+    public static final class CanceledCause extends CauseOfInterruption {
```

Comment:
Consider having a more descriptive name, since some semi badly written statistics calculators are using the simpleName of the class as category display name.